### DISCUSSION TITLE
What do you think of TMap#asTRef

### DISCUSSION THREAD

**Discussion by regiskuckaertz** [zio/zio#5974] [Ideas]
Hello gang 👋 

I have a service that pools database connections. These connections are created on-the-fly and then reused by multiple components over their lifetime. For security reasons, db credentials expire after a few hours, so we must take care of renewing these connections. So in my module implementation I have a cache:

```scala
conns: TMap[Role, TPromise[DatabaseError, Database]]
```

When connections are issued, we look into the map and if nothing is found, create a new promise and an action to fetch credentials, yaddi yadda. The consumer receives a `Ref` filled with that database and which will be updated whenever the connection is recycled. I thought that was a bit redundant and wanted to suggest adding:

```scala
trait TMap[K, V]:
  def asTRef(k: K): STM[TRef[V]]
```

that would zoom in on a specific element in the map and automatically give read/write access to it. Obviously my use case is quite specific, so this may not be useful at all.

**Comment by adamgfraser** [comment]
I think conceptually this makes sense. It is basically an optic that "zooms in" to the value at a particular key. I think there are some implementation issues because a `TRef` is currently always at least backed by a single `TRef.Atomic` and this is the backbone of transactional operations within STM, so exposing a `TRef` that did not satisfy that property would require significant thought.

Have you looked at using the new `ZPool` data type for this?

**Reply by regiskuckaertz** [reply]
Wow, I did not see that. I see, that would simplify things quite a bit.

Thank you for the tip 🙌
